itests/hive-unit/src/test/java/org/apache/hive/jdbc/TestJdbcTimeout.java

```diff
@@ -0,0 +1,87 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hive.jdbc;
+
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hdfs.DFSUtil;
+import org.apache.hadoop.hdfs.HAUtil;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.conf.HiveConf.ConfVars;
+import org.apache.hive.jdbc.miniHS2.MiniHS2;
+import org.apache.hive.service.cli.HiveSQLException;
+import org.apache.hive.service.cli.session.HiveSessionHook;
+import org.apache.hive.service.cli.session.HiveSessionHookContext;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.junit.*;
+
+import java.sql.*;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+public class TestJdbcTimeout {
+
+  private static MiniHS2 miniHS2 = null;
+
+  @BeforeClass
+  public static void beforeTest() throws Exception {
+    Class.forName(MiniHS2.getJdbcDriverName());
+    HiveConf conf = new HiveConf();
+    conf.setBoolVar(ConfVars.HIVE_SUPPORT_CONCURRENCY, false);
+    miniHS2 = new MiniHS2.Builder().withConf(conf).build();
+    miniHS2.start(new HashMap<>());
+  }
+
+  @AfterClass
+  public static void afterTest() throws Exception {
+    if (miniHS2 != null && miniHS2.isStarted()) {
+      miniHS2.stop();
+    }
+  }
+
+  @Test
+  public void testConnection() throws Exception {
+    AtomicInteger goodCounter = new AtomicInteger(0);
+    AtomicInteger badCounter = new AtomicInteger(0);
+    DriverManager.setLoginTimeout(2);
+    String url = miniHS2.getJdbcURL();
+    Runnable run = () -> {
+      try (Connection conn = DriverManager.getConnection(url)) {
+        goodCounter.incrementAndGet();
+        try (Statement stmt = conn.createStatement()) {
+          stmt.execute("SELECT reflect('java.lang.Thread', 'sleep', 3000L)");
+        }
+      } catch (Exception exception) {
+        badCounter.incrementAndGet();
+      }
+    };
+    Thread conn1 = new Thread(run);
+    Thread conn2 = new Thread(run);
+    conn1.start();
+    conn2.start();
+    assertEquals(goodCounter.get(), 1);
```

Comment:
Couldn't this become flaky, if the `getConnection` is slow?

Reply:
WIP: I can not compile Hive on M1 machine, just push and finding a x86 machine...

Reply:
I'm sorry I can not compile and run the test even on an x86 machine, the IDEA is stuck on building hive-exec module... and I simplified the test to check the variable initialization only, please let me know if this is sufficient or not.

Reply:
I would like to see some test where we make sure that no future change will break the `loginTimeout`, `socketTimeout` behaviour.

Not sure about your env, but we sometime create virtual machines and run our tests there to have a linux env. This might help you too if you have access to some env.